NExtends/Expressions/ExpressionHelper.cs

```diff
@@ -0,0 +1,235 @@
+
+		public class ParameterReplacementVisitor : ExpressionVisitor
+		{
+			private readonly Expression _oldParameter;
+			private readonly Expression _newParameter;
+			private readonly Dictionary<string, string> _propertyRenaming;
+
+			public ParameterReplacementVisitor(ParameterExpression oldParameter, ParameterExpression newParameter, Dictionary<string, string> propertyRenaming)
+			{
+				_oldParameter = oldParameter;
+				_newParameter = newParameter;
+				_propertyRenaming = propertyRenaming;
+			}
+			protected override Expression VisitParameter(ParameterExpression node)
+			{
+				if (_oldParameter == node)
+				{
+					return _newParameter;
+				}
+				else
+				{
+					return base.VisitParameter(node);
+				}
+			}
+			protected override Expression VisitMember(MemberExpression node)
+			{
+				if (node.Member.DeclaringType == _oldParameter.Type)
+				{
+					var propertyName = node.Member.Name;
+					if (_propertyRenaming.ContainsKey(propertyName))
+					{
+						propertyName = _propertyRenaming[propertyName];
+					}
+					var propertyInfo = _newParameter.Type.GetProperty(propertyName);
+					return Expression.MakeMemberAccess(_newParameter, propertyInfo);
+				}
+				else
+				{
+					return base.VisitMember(node);
+				}
+			}
+		}
+		public static Expression BuildAny<TSource>(Expression accessor, Expression predicate)
+		{
+			return BuildAny(typeof(TSource), accessor, predicate);
+		}
+		public static Expression BuildAny(Type tSource, Expression accessor, Expression predicate)
+		{
+			return BuildEnumerableMethod(tSource, accessor, predicate, "Any");
+		}
+
+		public static Expression BuildWhere<TSource>(Expression accessor, Expression predicate)
+		{
+			return BuildWhere(typeof(TSource), accessor, predicate);
+		}
+		public static Expression BuildWhere(Type tSource, Expression accessor, Expression predicate)
+		{
+			return BuildEnumerableMethod(tSource, accessor, predicate, "Where");
+		}
+
+		private static Expression BuildEnumerableMethod(Type tSource, Expression accessor, Expression predicate, string method)
+		{
+			var overload = typeof(Enumerable).GetMethods()
+									  .Single(mi => mi.Name == method && mi.GetParameters().Count() == 2).MakeGenericMethod(tSource);
+
+			var call = Expression.Call(
+				overload,
+				accessor,
+				predicate);
+
+			return call;
+		}
+
+		public static bool HasSubexpressionsOfType<TTest>(Expression original)
+		{
+			var visitor = new HasTypeVisitor<TTest>();
+			return visitor.VisitRoot(original);
+		}
+
+		private class HasTypeVisitor<TKeep> : ExpressionVisitor
+		{
+			private Type _toKeep;
+			private bool _HasType = false;
+			public HasTypeVisitor()
+			{
+				_toKeep = typeof(TKeep);
+			}
+			public bool VisitRoot(Expression node)
+			{
+				var visited = Visit(node);
+				return _HasType;
+			}
+			public override Expression Visit(Expression node)
+			{
+				if (_HasType) { return node; }
+				if (node != null && node.Type == _toKeep)
+				{
+					_HasType = _HasType | true;
```

Comment:
![BLOCKER](https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER') Correct this '|' to '||'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2178)